---
title: "Count features"
author: "Charles Plessy"
date: "02/07/2023"
output: 
  html_document: 
    keep_md: yes
params:
  alnFile: "/absolute/path/to/your/file"
  matchType: "match_part or match depending on GFF file format"
  prefix: "PrefixForOutputFileNames"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(progress = TRUE, verbose = TRUE, root.dir = system2("pwd", stdout = TRUE))
```

```{r load_libraries}
library("GenomicBreaks") |> suppressPackageStartupMessages()
library("ggplot2")       |> suppressPackageStartupMessages()
```

How to use this Rmarkdown template:

It is essential to set `R_LIBS_USER` to an empty value, otherwise `R` will
attempt to load packages from the user's library, which were compiled in a
different environment than the GenomicBreaks Singularity image.

Example command

```
R_LIBS_USER='' R -e 'rmarkdown::render("thisTemplate.Rmd", output_file = "./outFile.html", params = list(alnFile = "/path/do/dir/with/alignment/files"))'
```

Load data
---------

```{r load_data}
# Display parameters
params

# Load the alignment in a GBreaks object
gb <-load_genomic_breaks(params$alnFile, type = params$matchType)
```

Coalesce contigs
----------------

```{r coalesce_contigs}
coa <- coalesce_contigs(gb)
```

### Extract information

We divide the genome into four categories: _isolated alignments_, _breakpoint
regions_, _colinear alignments_ and _bridge regions_.

_Colinear alignments_ are defined by the colinearity relationship computed
in `?flagColinearAlignments`.  _Bridge regions_ separate alignments that are
colinear to each other.  _Isolated alignments_ have no colinear counterparts
and _breakpoint regions_ are the remaining intervals.  _Colinear regions_ are
the union of _colinear alignments_ and _bridge regions_.

As Bioconductor's `gaps()` function returns also the unaligned sequences between
the start/end of chromosome and the first/last block, we use the `?cleanGaps`
function that removes them before returning the object.
   
```{r clean_gaps}
isol <- gb[gb %in% coa]
coli <- gb[!gb %in% coa]
bri  <- bridgeRegions(gb)
brk  <- cleanGaps(coa)
```

```{r plot_widths, dev = c('svg', 'png')}
# Need to wrap in a function to handle empty GRanges objects...
width2df <- function(what, gr) {
  if (length(gr) == 0) {
    data.frame(what = what, width = 0)
  } else { 
    data.frame(what = what, width = width(gr))
  }
}

rbind(
  width2df(what = "isol_aln", gr = isol),
  width2df(what = "breakpoint_regions",   gr = brk),
  width2df(what = "colinear_aln", gr = coli),
  width2df(what = "colinear_region", gr = coa),
  width2df(what = "bridge", gr = bri)
) |> ggplot() +
  aes(width) +
  geom_histogram() +
  scale_x_log10() +
  facet_wrap(~what, ncol = 1)
```

Inversions
----------

```{r study_inversions}
coa <- flagInversions(coa)
sum(coa$inv)
showInversions(coa) |> head(11)
if (sum(coa$inv) > 10) {
  # Histogram of the widths
  ggplot(data.frame(width = width(filterInversions(coa)))) +
    aes(width) +
    geom_histogram() +
    scale_x_log10()
}
```

Translocations
--------------

Patterns that can be described as translocations in the target genome.

```{r study_translocations}
coa <- flagTranslocations(coa)
sum(coa$tra)
showTranslocations(coa) |> head(11)

if (sum(coa$tra) > 10) {
  # Histogram of the widths
  ggplot(data.frame(width = width(filterTranslocations(coa)))) +
    aes(width) +
    geom_histogram() +
    scale_x_log10()
}
```


Calculate numbers and prepare them for export in a YAML file
------------------------------------------------------------

```{r count_features}
report <- list()

report[["aligned_length_total"]]  <- sum(gb$aLength)
report[["aligned_target_total"]]  <- sum(width(gb))
report[["aligned_query_total"]]   <- sum(width(gb$query))
report[["aligned_ranges_number"]] <- length(gb)

report[["colinear_target_total"]] <- sum(width(coa))
report[["colinear_query_total"]]  <- sum(width(coa$query))
report[["colinear_ranges_number"]] <- length(coa)

report[["percent_similarity"]] <- 1 - sum(gb$matches) / sum(gb$aLength)

report[["number_of_inversions"]]       <- sum(coa$inv)
report[["number_of_transpositionsl"]]  <- sum(coa$tra)

report[["synteny_index"]]              <- synteny_index(gb)
report[["correlation_index"]]          <- correlation_index(gb)
report[["GOC_vicinity_4"]]             <- GOC(gb, vicinity = 4) # Default as of today
report[["strand_randomisation_index"]] <- strand_randomisation_index(gb)
```

Export the results to a YAML file.

```{r export_results}
yaml <- yaml::as.yaml(report)
yaml::write_yaml(yaml, paste0(params$prefix, ".yaml"))
cat(yaml)
```
